book/src/proposals/booting-from-snapshot.md

```diff
@@ -0,0 +1,316 @@
+For each included validator, the set of votes will be the votes on the fork
+that achieves the highest lockout on `snapshot_root` (up to
+`MAX_LOCKOUT_HISTORY`). The complete vote transaction will be included so that
+the signature is verifiable.  If available, a vote for `snapshot_root` must be
+included to verify `snapshot_root`'s bank hash.
+
+
+## Boot Procedure:
+
+1) The validator gets a snapshot to boot from. This can come from any source,
+as it will be verified before being vote upon.
+
+2) The validator rejects any snapshot where `snapshot_root` <= the largest root
+recorded in blocktree.
+
+3) The validator rejects any snapshot where `tower.root` is not an ancestor of
+`snapshoot_root`.  If `tower.root` < `snapshot_root`, this can be done by
+checking if `tower.root` is contained in the snapshot's `ancestors` set.  If
+`tower.root` == `snapshot_root` then ancestry is proven. `tower.root` <=
+`snapshot_root` provided that the check in 2) passes and that the order of
+events when setting a new root is:
+
+    1) Write the root to tower
+    
+    2) Write the root to blocktree
+    
+    3) Generate snapshot for that root
+
+4) On startup, the validator checks that the current root in tower exists in
+blocktree. If not (there was a crash between 3-1 and 3-2), then rewrite the
+root to blocktree.
+
+5) On startup, the validator checks that `snapshot_votes` includes votes that
+prove threshold commitment from the trusted validators on `snapshot_root`. See
+the `Trusted Validator Set` section for details. If not, the validator panics
+and different snapshot must be obtained.
+
+6) On startup, the validator boots from the `snapshot_root`, then replays all
+descendant blocks of `snapshot_root` that exist in this validator's ledger,
+building banks which are then stored in an output `bank_forks`. This is done in
+`blocktree_processor.rs`. The root of this `bank_forks` is set to
+`snapshot_root`.
+
+7) On startup, the validator calculates the `first_votable_slot` from which it
+is not locked out for voting. Every validator persists its tower state and must
+consult this state in order to boot safely and resume from a snapshot without
+being slashed. From this tower state and the ancestry information embedded in
+the snapshot, a validator can derive which banks, are "safe" (See the
+`Determining Vote Safety From a Snapshot and Tower` section for more details)
+to vote for.
+
+8) Periodically send canary transactions to the cluster using the validator's
+local recent blockhash.
+
+9) Wait for the following criteria. While waiting, set a new root every time
+2/3 of the cluster's stake roots a bank. This allows the validator to prune its
+state and also calculate leader schedules as it moves across epochs. Even if
+the validator appears on the leader schedule, it does not produce blocks while
+waiting.
+
+   1) Observe a canary transaction in a bank that some threshold of trusted
```

Comment:
It seems to guarantee you are voting on the tip of a fork, either 

1) You see a canary get rooted and can trust all forks that descend from that root (this can't happen if more than 33% of the stake all crashed and is rebooting)
2) You need a canary for every fork

If you see a canary on one fork you know you are caught up on that fork. But if you don't have a canary for every fork, it's possible you then switch to another heavier fork that you are not at the tip of. Is that an issue?